rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,815 @@
+
+        # only the RasaFeatureCombiningLayer is present
+        return self._tf_layers["feature_combining_layer"].output_units
+
+    def _features_as_token_ids(
+        self, features: List[Union[tf.Tensor, tf.SparseTensor]]
+    ) -> Optional[tf.Tensor]:
+        """Creates dense labels (token IDs) used for negative sampling in MLM."""
+        # If there are dense features, we use them as labels - taking the first dense
+        # feature in the list, but any other dense feature would do the job.
+        for f in features:
+            if not isinstance(f, tf.SparseTensor):
+                return tf.stop_gradient(f)
+
+        # If no dense features are found, use a sparse feature but convert it into
+        # a dense one first.
+        for f in features:
+            if isinstance(f, tf.SparseTensor):
+                return tf.stop_gradient(self._tf_layers["sparse_to_dense_token_ids"](f))
+
+    def _create_mlm_tensors(
+        self,
+        sequence_features: List[Union[tf.Tensor, tf.SparseTensor]],
+        seq_sent_features: tf.Tensor,
+        mask_combined_sequence_sentence: tf.Tensor,
+        sentence_features_present: bool,
+        training: bool,
+    ) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+        """Produce helper variables for masked language modelling (only in training).
+        
+        The `token_ids` embeddings can be viewed as token-level labels/unique IDs of all
+        input tokens (to be used later in the MLM loss) because these embeddings aren't 
+        affected by dropout or masking and are effectively always unique for different 
+        input tokens (and same for the same tokens).
+        `token_ids` share the batch and sequence dimension with the combined sequence-
+        and sentence-level features, the last dimension is unimportant and mimics the 
+        first dense sequence-level feature in the list of features, or alternatively the
+        last dimension will have size 2 if there are only sparse sequence features
+        present.
+        """
+        token_ids = self._features_as_token_ids(sequence_features)
+
+        # Pad in the sequence dimension to match the shape of combined sequence- and
+        # sentence-level features. This means padding by 1 if sentence-level features
+        # are present (those effectively have sequence length of 1) and not padding
+        # otherwise.
+        if sentence_features_present:
+            token_ids = tf.pad(token_ids, [[0, 0], [0, 1], [0, 0]])
+
+        # mlm_boolean_mask has the same shape as mask_combined_sequence_sentence (i.e.
+        # as the tensor with all combined features), with True meaning tokens that are
+        # masked and False meaning tokens that aren't masked or that are fake
+        # (padded) tokens.
+        seq_sent_features, mlm_boolean_mask = self._tf_layers["mlm_input_mask"](
+            seq_sent_features, mask_combined_sequence_sentence, training
```

Comment:
shouldn't this be the mask of only sequence? I don't remember

Reply:
Actually, I'm a bit surprised too (and this isn't the 1st time I'm thinking about it). But it is the combined mask in the original code as well (there, `mask_sequence` is used for just sequence-level features mask and `mask`/`mask_text` is used for combined features mask). See [here](https://github.com/RasaHQ/rasa/blob/8759a5dff72fb8df5b1e25edfda37f9f3d8dcc49/rasa/utils/tensorflow/models.py#L1063). Personally, I think it may be a bug 😉

Reply:
yes, looks like a bug. Could you change to sequence mask. Could you also add this change to changelog

Reply:
Alright, I'm changing it to use only sequence-level features and the corresponding mask.